emukit/core/loop/user_function.py

```diff
@@ -47,3 +47,50 @@ def evaluate(self, inputs: np.ndarray) -> List[UserFunctionResult]:
         for x, y in zip(inputs, outputs):
             results.append(UserFunctionResult(x, y))
         return results
+
+class MultiSourceFunctionWrapper(UserFunction):
+    """
+    Wraps a list of python functions that each correspond to different information source.
+    """
+
+    def __init__(self, f: List, source_index: int=-1) -> None:
+        """
+        :param f: A list of python function that take in a 2d numpy ndarrays of inputs and return 2d numpy ndarrays
+                  of outputs.
+        :param source_index: An integer indicating which column of X contains the index of the information source.
+                             Default to the last dimension of the input.
+        """
+        self.f = f
+        self.source_index = source_index
+
+    def evaluate(self, inputs: np.ndarray) -> List[UserFunctionResult]:
+        """
+        Evaluates the python functions corresponding to the appropriate information source
+
+        :param inputs: A list of inputs to evaluate the function at with information source index appended as last column
+        :return: A list of function outputs
+        """
+
+        if inputs.ndim != 2:
+            raise ValueError("User function should receive 2d array as an input, actual input dimensionality is {}".format(inputs.ndim))
+
+        n_sources = len(self.f)
+
+        # Run function for inputs at the first information source
```

Comment:
Can you explain why the code for the first source should be separate from the rest (sorry if I have already asked)? Also, if there is a valid reason, there is some code duplication here, might be worth writing a helper function

Reply:
I'll refactor, it was to avoid having to know the number of outputs before starting the loop but I'll do it a different way.